thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -2932,35 +2937,57 @@ class CBucketHandler : public CSimpleInterface, implements IInterface, implement
     {
         return spillBucket(critical);
     }
-    IRowStream * getNextBestRowStream()
+
+    void checkCompletedBuckets()
     {
         // NB: Called only once input has been read
         CriticalBlock b(spillCrit); // buckets can still be spilt
-        if (!bestReady)
-        {
-            // All non-spilled buckets in memory at this point in time are deduped
-            // -> set flag in all these buckets just in case they are spilled so that
-            //    when they need to be streamed back it's not necessary to dedup again
-            bestReady = true;
-            for (unsigned cur=0; cur<numBuckets; cur++)
-            {
-                if (!buckets[cur]->isSpilt())
-                    buckets[cur]->setRowsInBucketDeduped();
-            }
+        // All non-spilled buckets in memory at this point in time are fully deduped
+        // -> set flag in all these buckets just in case they are spilled so that
+        //    if they need to be streamed back it's not necessary to dedup again
+        for (unsigned cur=0; cur<numBuckets; cur++)
+        {
+            CBucket &bucket = *buckets[cur];
+            if (bucket.isSpilt())
+                bucket.close(); // close stream now, to flush rows out in write streams, so ready to be read
+            else
+                bucket.setCompleted();
         }
+    }
+/*
+ * NB: getNextBestRowStream() is only used when BEST involved
+ * It is called once all input rows have been consumed.
+ *
+ * Returns: a stream of next available unspilt, or spilt buckets
+ * that were marked complete (setCompleted())
+ */
+    IRowStream * getNextBestRowStream()
+    {
+        // NB: Called only once input has been read
+        CriticalBlock b(spillCrit); // buckets can still be spilt
         while (nextBestBucket < numBuckets)
         {
             CBucket *bucket = buckets[nextBestBucket++];
+
+            /* JCSMORE - It would be better to prioritize non-spilt buckets first
+             * So that those memory consuming buckets are consumed 1st.
+             */
             if (bucket->isSpilt())
             {
-                if (bucket->areRowsInBucketDeduped())
+                if (bucket->isCompleted())
                 {
+                    /* NB: getSpillRowStream() will be empty, because this bucket completed
+                     * And key stream are whole rows because BEST does not extract key.
+                     */
                     rowcount_t count; // unused
-                    return bucket->getSpillRowStream(&count);
+                    return bucket->getSpillKeyStream(&count);
```

Comment:
NB: this was previously reading the wrong stream, meaning that it would have lost results.
i.e. if a bucket spilt after being marked complete.